Sources/NIOExtras/LengthFieldPrepender.swift

```diff
@@ -0,0 +1,97 @@
+/// For example, if you received a packet with a 3 byte length:
+///     +---+-----+
+///     | A | BCD | ('A' contains 0x0003)
+///     +---+-----+
+///
+/// Given that the specified header length is 1 byte, there would be a byte appended which contains the number 3
+/// This initial appended byte is called the 'length field'.
+///
+public final class LengthFieldPrepender: MessageToByteEncoder {
+    
+    ///
+    /// An enumeration to describe the length of a piece of data in bytes.
+    /// It is contained to lengths that can be converted to integer types.
+    ///
+    public enum ByteLength {
+        case one
+        case two
+        case four
+        case eight
+       
+        fileprivate var length: Int {
+
+            switch self {
+            case .one:
+                return 1
+            case .two:
+                return 2
+            case .four:
+                return 4
+            case .eight:
+                return 8
+            }
+        }
+    }
+
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutbondOut = ByteBuffer
+
+    private let lengthFieldLength: ByteLength
+    private let lengthFieldEndianness: Endianness
+
+    /// Create `LengthFieldPrepender` with a given length field length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: ByteLength, lengthFieldEndianness: Endianness = .big) {
+        
+        // The value contained in the length field must be able to be represented by an integer type on the platform.
+        // ie. .eight == 64bit which would not fit into the Int type on a 32bit platform.
+        precondition(lengthFieldLength.length <= Int.bitWidth/8)
+        
+        self.lengthFieldLength = lengthFieldLength
+        self.lengthFieldEndianness = lengthFieldEndianness
+    }
+    
+    public func allocateOutBuffer(ctx: ChannelHandlerContext, data: OutboundIn) throws -> ByteBuffer {
+        return ctx.channel.allocator.buffer(capacity: self.lengthFieldLength.length + data.readableBytes)
+    }
+
+    public func encode(ctx: ChannelHandlerContext, data: OutboundIn, out: inout ByteBuffer) throws {
+        
+        switch self.lengthFieldLength {
+        case .one:
+            out.write(integer: UInt8(data.readableBytes), endianness: self.lengthFieldEndianness)
+        case .two:
+            out.write(integer: UInt16(data.readableBytes), endianness: self.lengthFieldEndianness)
+        case .four:
+            out.write(integer: UInt32(data.readableBytes), endianness: self.lengthFieldEndianness)
+        case .eight:
+            out.write(integer: UInt64(data.readableBytes), endianness: self.lengthFieldEndianness)
+        }
+        
+        out.write(bytes: data.readableBytesView)
```

Comment:
This passes the unit tests and works in my little test program, but I am not sure about the efficiency of writing using the byte view. If there is a more efficient way, would you mind advising so that I can take a look at a preferable way to achieve the same goal. Thanks.

Reply:
`out.write(buffer: data)` would be better here. However that will still do a copy that we don’t need.
If you make the whole class a `ChannelOutboundHandler` and in `func write` you first write the length and then just forward the incoming write, you can do this whole thing without allocating and without copying and that’ll be much more efficient.

NIO is smart enough that if you do

```
ctx.write(self.wrapOutboundOut(self.lengthBuffer), promise: nil)
ctx.write(data, promise: promise)
```

Both these writes will happen in one system call :)

Reply:
Perfect, exactly what I was looking for. Thanks. Will do.

Reply:
Okay, all done. Thanks for the tip.